### DISCUSSION TITLE
InlayHints default setting

### DISCUSSION THREAD

**Discussion by poyaoc97** [clangd/clangd#986] [General]
Hi all,
InlayHints seems to be on by default after https://github.com/llvm/llvm-project/commit/7c19fdd59939249c23384f0900d49aab4a5f0695 ([this line](https://github.com/llvm/llvm-project/blob/7c19fdd59939249c23384f0900d49aab4a5f0695/clang-tools-extra/clangd/tool/ClangdMain.cpp#L334) and [this line](https://github.com/llvm/llvm-project/blob/7c19fdd59939249c23384f0900d49aab4a5f0695/clang-tools-extra/clangd/ClangdLSPServer.h#L67) specifically), I wonder if this is intentional?

At first sight, I thought this was some new vscode voodoo😅, but then I figure it could be clangd (it was), which I rebuilt from ToT yesterday. This feature is cool tho, thanks for the effort! It will definitely come in handy someday just not today.

For future reference, I managed to turn it off by adding `--inlay-hints=false` in my vscode settings:
```json
"clangd.arguments": [
    "--log=verbose",
    "-j=4",
    "--background-index",
    "--clang-tidy",
    "--fallback-style=LLVM",
    "--header-insertion=iwyu",
    "--inlay-hints=false"
  ],
```

**Comment by sam-mccall** [comment]
Hi, thanks for asking and sorry if this was an unexpected surprise!

Yes, our plan is to turn this on by default on the server for clangd 14, and the release process will start in a few weeks.

If you have feedback on why you'd rather not have the feature enabled, we'd appreciate it - maybe we can use it to make it more useful/less annoying, or maybe we should reconsider making it default. It's hard to get feedback on this stuff before releasing the feature!

Sorry to be extra annoying but we actually plan to drop the `--inlay-hints` flag too. The best way to disable the feature is to put this in your [config.yaml](https://clangd.llvm.org/config):
```
InlayHints:
  Enabled: No
```

**Reply by tickley** [reply]
Hello chans, why can't I find the file config.yaml in my document? I have been troubled by the Mosaic prompt and can't find the file. use

"clangd.arguments": [
"--log=verbose",
"-j=4",
"--background-index",
"--clang-tidy",
"--fallback-style=LLVM",
"--header-insertion=iwyu",
"--inlay-hints=false"
],

 this won't turn off either

**Comment by oblitum** [comment]
A client might provide the option for displaying inlay hints or not, regardless whether the feature is hard-coded on the server. coc-clangd has this option for example. Another option that it provides is to set the separator character, but sadly, this client separator seems to be added on top of the server's hard-coded one (`:`), and that's indeed bad. Isn't there any flag to not have these server hard-coded separators? I wish to configure them on the client instead, or have a server option, or maybe this is a client issue.

cc @fannheyward.

**Comment by oblitum** [comment]
Scratch that, it seems `:` is just for label separation, not hints separation, but often you may not have labels in the hints at all, say if you do, `auto x = 0, y = 0;`, you won't have labels, but the `:` will still be present, so they're confused with hint separation. I think it would be better behavior to just remove `:` when a hint hunk doesn't have a label.

**Comment by oblitum** [comment]
IMHO, current default state of inlay hints isn't great, as in real code it's often mostly duplicating what's in the code already, and adding labels on top (`ParameterNames: Yes`), due to that I'm disabling them by default (`ParameterNames: No`), solely leaving `DeducedTypes: Yes`, which is usable (although there's no use for labels, and hence `:`, in this mode!).

![img-2022-06-07-143431](https://user-images.githubusercontent.com/1269815/172446700-c09b7d58-cce4-4e4c-9341-49a9f7bed289.png)

**Reply by HighCommander4** [reply]
I don't really understand what's going on in this screenshot.

The annotations at the ends of the line... is that how this editor visualizes clangd's inlay hints? Or is that a separate but similar feature provided by this editor?

**Comment by oblitum** [comment]
@HighCommander4 it's how clangd's inlay hints are visualized, through the editor (nvim + coc.nvim + coc-clangd).

**Comment by oblitum** [comment]
TBH, I'm not sure how implementation is behaving, coc-rust-analyzer has more sane presentation, coc-clangd not so much, but I can't state with confidence whether that's due to clangd inlay hints data or how coc-clangd presents that.

Vim and NeoVim have a limitation regarding inserting inlay hints inside the code, it can only present them besides, so, that code duplication for purposes of pointing what's being labeled might be solely a client issue/limitation.

https://github.com/neovim/neovim/pull/9496

**Comment by oblitum** [comment]
Thanks for the question, now I see, this bad experience must all be due to this editor limitation. Thinking how VSCode works (I don't have it installed, but I know how hints are implemented with tiny text in these editors), it all makes sense now, the presence of `:` even when there's no labels, etc, it's all designed with intermingling inlay hints in mind... they're supposed to always be directly besides each target, not at the end of line.

**Reply by oblitum** [reply]
The editor limitation topic is settled for Vim as it's supporting inlay hints in middle text already, it's a matter of plugins picking the feature up now. NeoVim as of this date still can only show it at the end of line, which can get very confusing.
